### DISCUSSION TITLE
Database class save and read issues

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#136] [Design & Development]
Working on the getting started tutorial caused me to quickly recognize two problems we need to deal with.

1.  In this tutorial I have the user download data from IRIS with obspy's web service function and immediately save the results to wf_TimeSeries.  Works really slick because it avoids the giant seed files like those we are dealing with for the usarray data.   Shows clearly, however, once again why right now the just in time model for data storage is not workable as it takes a solid 30 minutes to download just 3 usarray event files smaller than those I downloaded before.  What came up was that what we should want to show, which is this block of code:
```
strm=client.get_waveforms(
            starttime=UTCDateTime(stime),
            endtime=UTCDateTime(etime),
            network='TA',
            channel='BH?',
            location='*'
        )
        for d in strm:
            d_mspass=Trace2TimeSeries(d)
            print('Saving data for sta=',d_mspass['sta'],' and chan=',d_mspass['chan'])
            db.save_data(d_mspass)
```
runs fine but produces useless wf_TimeSeries entries.   A typical document saved that way from the draft tutorial looks like this:
```
{'_id': ObjectId('601993a7e4c0ed39f9898a26'),
 'calib': 1.0,
 'delta': 0.025,
 'npts': 59081,
 'sampling_rate': 40.0,
 'starttime': 1299825273.0,
 'time_standard': 'UTC',
 'gridfs_id': ObjectId('601993a7e4c0ed39f9898a27'),
 'storage_mode': 'gridfs'}
```
The data there are useless because there is nothing to link them to anything in site, channel, or source.  Source is a different issue that was a mistake in my script, but the issue is channel data.   The problem is the save_data function defaults to *include_undefined=False*.  As a result the net, sta, chan, and loc attributes are dropped in the save.   To fix this we need to do one of two things: (1) add net, sta, chan, and loc to the wf_TimeSeries schema as optional parameters, or (2) make the default for include_undefined True.  I would argue actually that we should do both.  The biggest user base for MsPASS will be people getting data from IRIS and not carrying net, sta, chan, and loc along like this will cause endless emails of complaints if left that way.  I actually think defaulting to include_undefined should default to true for the same reason.  Better to save junk than to lose critical data by mistake, especially since pymongo is pretty promiscuous on type.  
2. I think it would be a good convenience to the user to have *read_data* accept either an ObjectId or a doc from which would extract the ObjectId.  The difference is  only 2 or 3 lines of code.  For a beginner that is one less concept to have to grasp.   The counter argument is allowing a parameter to specify two drastically different data types is bad form and potentially causes more confusion than it fixes. 

In summary, some change for item 1 is essential.  Item 2 is a topic for discussion on which I am not as sure.

**Comment by pavlis** [comment]
I may have put the blame in the wrong place.  The wf_TimeSeries schema, I think, should have net, sta, chan, and loc added, BUT there adding *include_undefined=True" did not fix the problem.  It appears our Trace2TimeSeries converter is not working right as it should have transferred those attributes to the TimeSeries it creates.  I'm going to have to do some debugging here. 

This issue does not really change the need for a discussion of the two points above.

**Reply by pavlis** [reply]
That guess was wrong.  There was nothing wrong with Trace2TimeSeries.  The problem is in the save_data method and how it interacts with the schema class.   Debugging this showed this is the problem:
1.  Within *save_data* is this line:
```
if isinstance(mspass_object, TimeSeries):
     update_metadata_def = schema.TimeSeries
```
2.  That is used while looping over Metadata keys here:
```
for k in copied_metadata:
                if k in exclude_keys:
                    continue
                if update_metadata_def.is_defined(k):
                    if update_metadata_def.readonly(k):
                        continue
                    if not isinstance(copied_metadata[k], update_metadata_def.type(k)):
                        try:
                            # The following convert the actual value in a dict to a required type.
                            # This is because the return of type() is the class reference.
                            insert_dict[k] = update_metadata_def.type(
                                k)(copied_metadata[k])
                        except Exception as err:
                            raise MsPASSError('Failure attempting to convert key {} from {} to {}'.format(
                                k, copied_metadata[k], update_metadata_def.type(k)), 'Fatal') from err
                    else:
                        insert_dict[k] = copied_metadata[k]
```
This creates the problem because in the Metadata schema net, sta, chan, and loc are tagged as "readonly: True".   

What I don't understand here is why the Metadata schema is overlaid on that for TimeSeries. TimeSeries does not even define net, sta, chan, and loc so seems to me they should be undefined there.   

Something is definitely wrong and I'm not sure how to fix it.  

Here is a script you can run that should allow you to completely duplicate this problem:
```
from mspasspy.db.database import Database
from mspasspy.db.client import Client as DBClient
dbclient=DBClient()
db=Database(dbclient,'test')
from obspy import UTCDateTime
from mspasspy.util.converter import Trace2TimeSeries

# We have to redefine the client from obspy to use their so called bulk downloader
from obspy.clients.fdsn import RoutingClient
client = RoutingClient("iris-federator")

t0=UTCDateTime('2011-03-11T05:46:24.0')
origin_time=t0.timestamp
start_offset=763.0-60.0
end_offset=1400.0+2*60.0
stime=origin_time+start_offset
etime=origin_time+end_offset

strm=client.get_waveforms(
            starttime=UTCDateTime(stime),
            endtime=UTCDateTime(etime),
            network='TA',
            channel='BHZ',
            location='*'
        )
ens=list()
for tr in strm:
    d=Trace2TimeSeries(tr)
    ens.append(d)
print('number converted=',len(ens))
db.save_data(ens[0],include_undefined=True)
```
You can see what happens only by using this with a debugger and stepping through save_data

**Comment by wangyinz** [comment]
yeah, I was looking at it, too, and found the same problem. The reason Metadata schema is used here is that we are saving a data object, so we need to look at its metadata to know what is defined and how they should be saved.

The problem here is the disconnection in handling readonly. Those site/channel information are readonly because in Metadata definition, they belong to the site or channel collection. Thus, shouldn't be updated because those collections should never change or at least not in a wf update. Such behavior is meaningless right now because the `save_data` method does not support updating other collections than the targeting wf one anyway. I guess the question is, what shall we do with keys that belong to a different collection. Maybe we should change that section of code into something like this:
```python
                if k in exclude_keys:
                    continue
                if update_metadata_def.is_defined(k) and update_metadata_def.writeable(k):
                    if not isinstance(copied_metadata[k], update_metadata_def.type(k)):
                        try:
                            # The following convert the actual value in a dict to a required type.
                            # This is because the return of type() is the class reference.
                            insert_dict[k] = update_metadata_def.type(
                                k)(copied_metadata[k])
                        except Exception as err:
                            raise MsPASSError('Failure attempting to convert key {} from {} to {}'.format(
                                k, copied_metadata[k], update_metadata_def.type(k)), 'Fatal') from err
                    else:
                        insert_dict[k] = copied_metadata[k]
                elif include_undefined:
                    insert_dict[k] = copied_metadata[k]
```

**Reply by wangyinz** [reply]
Well, actually that's still not right... we don't want to override `readonly` with the `include_undefined` argument.

**Reply by wangyinz** [reply]
Well, actually we encountered the problem here mostly because the `mspass.yaml` defines a normalized schema, but here in this example, you really want to use an embedded schema. Maybe the solution is to just use a different schema file.

**Reply by wangyinz** [reply]
Probably a schema like this should work:
```yaml
Database:
  wf_TimeSeries:
    default: wf
    data_type: TimeSeries
    schema: 
      chan:
        type: string
        concept: channel name (e.g. HHZ, BHE, etc.) - normally a SEED channel code
        aliases: [channel, KCMPNM, channel.chan, wfdisc.chan]
      <<: &wf_TimeSeries
        _id:
          type: ObjectID
          concept: ObjectId used to define a data object
        npts:
          type: int
          concept: Number of data samples
          aliases: [nsamp, wfdisc.nsamp]
        delta:
          type: double
          concept: Data sample interval in seconds
          aliases: [dt]
        sampling_rate:
          type: double
          concept: Data sampling frequency in Hz=1/s
          optional: true
        starttime:
          type: double
          concept: Time of first sample of data (epoch time or relative to some other time mark)
          aliases: [t0, time]
        starttime_shift:
          type: double
          concept:  Time shift applied to define relative time of 0.0
          aliases:  t0shift
        utc_convertible:
          type:  bool
          concept:  When true starttime_shift can be used to convert relative time to UTC.
        time_standard:
          type: string
          concept: Defines time standard for meaning of t0 (default should be UTC)
        calib:
          type: double
          concept: Nominal conversion factor for changing sample data to ground motion units.
        storage_mode:
          type: string
          concept: The storage mode of the saved data ('file', 'url' or 'gridfs')
        dir:
          type: string
          concept: Directory path to an external file (always used with dfile)
          aliases: [file.dir, wfdisc.dir, wfprocess.dir]
          optional: true
        dfile:
          type: string
          concept: External data file name.
          aliases: [file.dfile, wfdisc.dfile, wfprocess.dfile]
          optional: true
        foff:
          type: int
          concept: Offset in bytes from beginning of a file to first data sample.
          aliases: [file.foff, wfdisc.foff, wfprocess.foff]
          optional: true
        url:
          type: string
          concept: Url to external data file.
          optional: true
        net:
          type: string
          concept: network code (net component of SEED net:sta:chan)
          aliases: [network, site.net, channel.net, wfdisc.net, site_net]
        sta:
          type: string
          concept: station code assigned to a spot on Earth (sta component of SEED net:sta:chan)
          aliases: [station, site.sta, channel.sta, wfdisc.sta, KSTNM]
        loc:
          type: string
          concept: location code assigned to an instrument (loc component of SEED net:sta:chan)
          aliases: [location, site.loc, channel.loc]
          optional: true
        gridfs_id:
          type: ObjectID
          concept: The _id of data file in the fs.files collection.
          optional: true
        history_object_id:
          reference: history_object
          optional: true
        elog_id:
          type: list
          reference: elog
          optional: true

  wf_Seismogram:
    data_type: Seismogram
    schema:
      <<: *wf_TimeSeries
      tmatrix:
        type: list
        concept: Three-component data's transformation matrix
  
  history_object:
    schema:
      _id:
        type: string
        concept: UUID used to define an unique entry in history collection.
      nodedata:
        type: bytes
        concept: serialized content of ProcessingHistory.

  elog:
    schema:
      _id:
        type: ObjectID
        concept: ObjectID used to define an unique entry in elog collection.
      logdata:
        type: list
        concept: a list of LogData.
      wf_Seismogram_id:
        reference: wf_Seismogram
        optional: true
      wf_TimeSeries_id:
        reference: wf_TimeSeries
        optional: true
      gravestone:
        type: dict
        concept: a copy of the Metadata of a dead object.
        optional: true


Metadata:
  TimeSeries:
    schema:
      _id:
        collection: wf_TimeSeries
        readonly: true
      npts:
        collection: wf_TimeSeries
        readonly: false
      delta:
        collection: wf_TimeSeries
        readonly: false
      sampling_rate:
        collection: wf_TimeSeries
        readonly: false
      starttime:
        collection: wf_TimeSeries
        readonly: false
      time_standard:
        collection: wf_TimeSeries
        readonly: false
      calib:
        collection: wf_TimeSeries
        readonly: false
      net:
        collection: wf_TimeSeries
        readonly: false
      sta:
        collection: wf_TimeSeries
        readonly: false
      loc:
        collection: wf_TimeSeries
        readonly: false
      chan:
        collection: wf_TimeSeries
        readonly: false
      
  Seismogram:
    schema:
      _id:
        collection: wf_Seismogram
        readonly: true
      npts:
        collection: wf_Seismogram
        readonly: false
      delta:
        collection: wf_Seismogram
        readonly: false
      sampling_rate:
        collection: wf_Seismogram
        readonly: false
      starttime:
        collection: wf_Seismogram
        readonly: false
      time_standard:
        collection: wf_Seismogram
        readonly: false
      calib:
        collection: wf_Seismogram
        readonly: false
      tmatrix:
        collection: wf_Seismogram
        readonly: false
      net:
        collection: wf_Seismogram
        readonly: false
      sta:
        collection: wf_Seismogram
        readonly: false
      loc:
        collection: wf_Seismogram
        readonly: false 
```

**Comment by wangyinz** [comment]
I added that yaml file in pull request #138. Note that instead of adding attributes to `mspass.yaml`, I created the `mspass_lite.yaml` file that defines a flat namespace needed for the tutorial. I also added new arguments to `Database.__init__`, and you can now construct the db with
```
db = Database(client, 'dbtest', db_schema="mspass_lite.yaml", md_schema="mspass_lite.yaml")
```
to use the new flat schema. 

I did not make `include_undefined=True` as the default because the logic of our database API is that all the default behavior should be defined by the schema, and any additional arguments in the API are used to override such default when needed (i.e., 'exclude_keys', 'include_undefined', etc). If we always set `include_undefined=True`, it invalidates the point of having a schema. Also, the other reason that make me hesitate to dump everything into database is that I looked at the database record created with `include_undefined=True` and found it really messy. There are meaningless attributes such as
```
"_fdsnws_dataselect_url" : "http://service.iris.edu/fdsnws/dataselect/1/query"
"_format" : "MSEED",
"mseed" : { "dataquality" : "M", "number_of_records" : 375, "encoding" : "STEIM2", "byteorder" : ">", "record_length" : 512, "filesize" : 570880 },
"processing" : [ "ObsPy 1.2.1: trim(endtime=UTCDateTime(2010, 2, 27, 7, 45, 0, 19538)::fill_value=None::nearest_sample=True::pad=False::starttime=UTCDateTime(2010, 2, 27, 6, 45, 0, 19538))" ],
```
I am pretty sure we want to filter those out for any of the wf entries, and we have discussed similar issues before. 

Also, if we set `include_undefined=True` in `save_data`, we probably should do the same in `read_data` for consistency. Then, that creates another problem: attributes such as `storage_mode` and `foff` are not useful at all and sometimes misleading in the data object constructed from the read. So, probably we should just leave the argument as false.

**Reply by pavlis** [reply]
This working now but I note it required this to work:
```
    d=db.read_data(id,load_history=False)
```
I thought the standard default would be False, but it currently is true for read_data.  I can see an argument for making it True by default, but it could have unintended consequences because one model for turning on history is to test if ProcessingHistory is empty.  It will not be if this is the default.  I can see this either way, but I think False is more consistent with what we have done so far.  It will simplify most usage and in our documentation we will need to make it clear that all modules need to enable history when needed.

**Reply by wangyinz** [reply]
The reason it is set as true right now is that we can run into problems if a user reads a data that has history record in the database but did not load the history and then saves it back as an update but with some new history in it. Right now, what happens will be that the old history record will be replaced by the new one, which means some of the history is lost. I wasn't able to figure out the best practice for such a situation, because if we simply merge the history record, it can cause more problems such as whether there is an overlap between the two history and how to make sure the global history captures such as merge.

But still, `load_history=True` will lead to the problem you described above, and it doesn't entirely avoid the problem of losing history. Anyway, I can change that to false for now, but we need to figure out the right way to handle history soon.

**Reply by wangyinz** [reply]
#139 got that addressed.

**Comment by wangyinz** [comment]
I also made `read_data` accept a dictionary with `_id` defined (https://github.com/wangyinz/mspass/pull/138/commits/675c8f2d9875aa2069e05a291bd2ff2b686a3d42).

**Reply by pavlis** [reply]
Looks like you made a lot of progress on this.   I'm going to need to resolve changes I had to make to database.py with what you did here to get all of this tutorial working.   Recall I had to fix a problem with save_inventory where it was not handling different ways inventory could be constructed.  I could hack through this, but I could use some education on best practices this point that might be useful to others reading this.  

Quick summary of situation:  You made some fixes to several files and merged them into master.  Origin is now several commits ahead of where my current branch was forked.  My branch has a couple commits that will absolutely conflict with your changes (database.py).   What is the way you want me to merge these?   

I  would previously have done this:
make sure my working branch was clean
git checkout master
git pull origin master
git checkout myworkingbranch
git merge origin

That will likely generate conflicts I have to resolve, but I expect that.  Is that right or is should I do this with rebase?  If so, how?

**Reply by pavlis** [reply]
Different subject.  Guess you conclude the convenience of using the document dict for read outweighs the "bad form" of us old C/Fortran programmers who find it puzzling how a function can take multiple argument types.   I realized, however, on reflection that that type construct is very pythonic and in fact one of the things that make python such a useful glue language.   I also realized we do that already in many methods of Database already when using the *mspass_object* argument.

**Reply by wangyinz** [reply]
The cleanest way to merge will be an interactive rebase, but I guess that's too much hassle for you. I think it is fine to do a straight merge. I will squash those commits in the pull request so that we can still have a clean history.

**Comment by pavlis** [comment]
Something is still wrong with save_data.   I have this  cell in the getting_started tutorial:
```
from mspasspy.util.converter import Trace2TimeSeries

#db=Database(dbclient,'getting_started')
# We have to redefine the client from obspy to use their so called bulk downloader
from obspy.clients.fdsn import RoutingClient
client = RoutingClient("iris-federator")
# This uses the (admitted obscure) approach of a set container of keepers to reduce the download time
keepers=set()
keepers.add(4)
keepers.add(7)
keepers.add(10)

cursor=db.source.find({},no_cursor_timeout = True)
count=0
start_offset=763.0-4*60.0
end_offset=1400.0+10*60.0
i=0
for doc in cursor:
    if i in keepers: 
        origin_time=doc['time']
        # We need the ObjectId of the source to provide a cross reference to link our 
        # waveform data to the right source.  Better to do this now than later as association 
        # can be a big challenge
        id=doc['_id']
        
        print('Starting on event number',i,' with origin time=',UTCDateTime(origin_time))
        stime=origin_time+start_offset
        etime=origin_time+end_offset
        strm=client.get_waveforms(
            starttime=UTCDateTime(stime),
            endtime=UTCDateTime(etime),
            network='TA',
            channel='BH?',
            location='*'
        )
        # The output of get_waveforms is an obspy Stream object. Stream objects are iterable 
        # so we work through the group, converting each to a MsPASS TimeSeries, and then saving 
        # the results to the database.  
        for d in strm:
            d_mspass=Trace2TimeSeries(d)
            # Here is where we save the id linked to the source collection. 
            # We use a MsPASS convention that such metadata have name collection_id
            d_mspass['source_id'] = id
            #print('Saving data for sta=',d_mspass['sta'],' and chan=',d_mspass['chan'])
            db.save_data(d_mspass,include_undefined=True)   
            count += 1
    else:
        print('Skipping event number ',i,' with origin time=',UTCDateTime(origin_time))
    i += 1
print('Number of waveforms saved=',count)
```
A key point to note is this line:  `d_mspass['source_id'] = id` that is called immediately before calling save_data.  A typical document saved in that loop looks like this:
```
{'_id': ObjectId('601bda48cfc6b8d3a01cb12f'),
 '_fdsnws_dataselect_url': 'http://service.iris.edu/fdsnws/dataselect/1/query',
 '_format': 'MSEED',
 'calib': 1.0,
 'chan': 'BHE',
 'delta': 0.025,
 'endtime': 1299826750.0,
 'mseed': {'dataquality': 'M',
  'number_of_records': 31,
  'encoding': 'STEIM2',
  'byteorder': '>',
  'record_length': 4096,
  'filesize': 142506496},
 'net': 'TA',
 'npts': 59081,
 'sampling_rate': 40.0,
 'source_id': ObjectId('601bce167c394740be52b665'),
 'sta': '034A',
 'starttime': 1299825273.0,
 'time_standard': 'UTC',
 'utc_convertible': True,
 'gridfs_id': ObjectId('601bda48cfc6b8d3a01cb130'),
 'storage_mode': 'gridfs'}
```
The tutorial is using mspass_lite.   The key point is there is no "source_id" entry in the document saved to MongoDB even though it was explicitly set AND I used "include_undefined=True".   The later is not the issue because "source_id" is defined but the save_data is incorrectly dropping source_id in the save.  That is not the behavior it should ever have because source_id is a cross referencing id that is essential for the normalized data model.   I'm not sure how to fix this or I'd take it on.  I'm not yet familiar with enough the way the Schema object is supposed to work.

**Reply by pavlis** [reply]
Small addition:   I notice mspass_lite doesn't even have a definition of anything containing the string "source".  That means the "include_undefined=True" isn't working as it is supposed to either.  

Remind me why you think having include_undefined false by default is right?  I think we are asking for a of complaints by users where they are losing metadata attributes.

I wonder if we have a mismatch in viewpoint on how this should all work.  There are really only two ways I can foresee save_data being used:
1.  In the context as above where you are loading relatively raw data.   The distinction there is that the metadata may not be complete and will need later updates.   Such data may also be prone to more mismatched keys and the save should be totally promiscuous - i.e. include_undefined=True.   
2.  A save_data within a processing run is a very different thing for two reasons.  First,  there we can make more assumptions that the metadata are clean.  Second, history has to be handled very differently in that situation.  I don't think we are doing it right for that situation, but I have not yet dug deep enough to be sure.  

This api, I think, is trying to support  3rd situation that I suspect will be rare, but which is probably needed:  a metadata update only.   That situation, in fact, is already covered by the update_metadata method but we should advertise it as being for that purpose.   I think there may be a fundamental logic error in how save_data is handling all this.   I am wonder if we need to change the api of save_data to this:
```
def save_data(self,  mspass_object, storage_mode='gridfs', dfile=None, dir=None, 
   state='processing',exclude_keys=[], collection=None):
```
where the state argument is some mechanism to distinguish if the save is for data being loaded or data being saved during processing.  I'm suggesting a string with a fixed set of values but it could be done other ways.   The only other option I'm suggesting for now is "origin", "loading", or whatever word you want to use to signify this is newly loaded data that may have warts.  I have suggested removing "include_undefined" as it would not be necessary but would effectively be turned on automatically in the loading mode (the promiscuous mode tag I gave this idea above).  

In this model update_metadata would be viewed as a different thing.  It could be used in save_data but not necessarily.

The thing I don't think we are handling right is the history.  When this method is called with *state='processing'* and the data have history defined it needs to call the "map_as_saved" method of ProcessingHistory.

**Reply by wangyinz** [reply]
I am actually confused by the example you posted above. The `source_id` is in the document you posted:
```
...
 'sampling_rate': 40.0,
 'source_id': ObjectId('601bce167c394740be52b665'),
 'sta': '034A',
...
```
I also tested locally and can see all keys saved correctly with `include_undefined=True`. So, I am not sure what the problem is. Are you on the latest master branch?

I mentioned this a few posts above, but anyway, the current rationale of our database API is: all the default behavior should be defined by the schema, and any additional arguments in the API are used to override such default when needed. Then, if the users want to bypass the schema definition, they can simply call native pymongo API for any partial updates. Or, if the users want to edit the schema, they can do so by calling `Database.database_schema` or `Database.metadata_schema` directly. The schema class already supports adding entries or aliases. I just realized that I probably should also implement a erase method there.

**Reply by wangyinz** [reply]
The history part of the database API is really just a draft implementation. It actually got more flaws than that. For example, the one I mentioned a few post ago can completely break the idea of saving history:
> The reason it is set as true right now is that we can run into problems if a user reads a data that has history record in the database but did not load the history and then saves it back as an update but with some new history in it. Right now, what happens will be that the old history record will be replaced by the new one, which means some of the history is lost. I wasn't able to figure out the best practice for such a situation, because if we simply merge the history record, it can cause more problems such as whether there is an overlap between the two history and how to make sure the global history captures such as merge.

**Reply by wangyinz** [reply]
Another thought: if adding `include_undefined=True` to every call to the database is too tedious, we might consider adding a switch to the `Database` class, where we can specify something like `Database.strict = True`. What do you think?

**Reply by pavlis** [reply]
Hmmm - my bad on the source_id thing.  I need to dig into this to see what was actually going wrong.  I observed a problem somehow didn't see the key data in that output.

We should continue a discussion, however, about the issue of how save_data would be used.  I hadn't realized you were basically stalling on ProcessingHistory for now, BUT we must be sure our api doesn't "paint us into a corner".   Maybe I'm overly worried as python is pretty flexible - you can always just add a new optional argument.

**Reply by wangyinz** [reply]
Yeah, I think we will need to re-implement part of the object-level history after we have the global history anyway, so it is not that important to fix bugs in the object-level history at the moment.

**Reply by pavlis** [reply]
Turned out to be a dumb mistake and I was misled by an error in scanning that output.  Sorry for wasting your time.

BTW I plan to check in some changes later today.  Most are bug fixes to functions in ensembles.py BUT I found an bug I'm squashing right now in the Database method called load_seed_channel.   You will need to resolve those changes.  I'll try to remember to repeat this instruction in the pull request.
